project/Build.scala

```diff
@@ -223,36 +229,79 @@ object Build extends sbt.Build {
       )
   ) dependsOn (state_api % "test->test; provided", dsl % "provided")
 
-  lazy val services = Project(
-    id = "gearpump-services",
-    base = file("services"),
-    settings = commonSettings  ++ 
-      Seq(
-        libraryDependencies ++= Seq(
-          "io.spray" %%  "spray-testkit"   % sprayVersion % "test",
-          "io.spray" %%  "spray-httpx"     % sprayVersion,
-          "io.spray" %%  "spray-client"    % sprayVersion,
-          "io.spray" %%  "spray-json"    % sprayJsonVersion,
-          "com.wandoulabs.akka" %% "spray-websocket" % sprayWebSocketsVersion
-            exclude("com.typesafe.akka", "akka-actor_2.11"),
-          "org.webjars" % "angularjs" % "1.4.3", // We stick on 1.3.x until angular-ui-select works
-          "org.webjars" % "bootstrap" % "3.3.5",
-          "com.lihaoyi" %% "upickle" % upickleVersion,
-          "org.webjars" % "d3js" % "3.5.5",
-          "org.webjars" % "momentjs" % "2.10.3",
-          "org.webjars.bower" % "angular-loading-bar" % "0.8.0",
-          "org.webjars.bower" % "angular-smart-table" % "2.1.1",
-          "org.webjars.bower" % "angular-motion" % "0.4.2",
-          "org.webjars.bower" % "bootstrap-additions" % "0.3.1",
-          "org.webjars.bower" % "angular-strap" % "2.3.1",
-          "org.webjars.bower" % "angular-ui-select" % "0.12.0",
-          "org.webjars" % "select2" % "3.5.2",
-          "org.webjars" % "select2-bootstrap-css" % "1.4.6",
-          "org.webjars.bower" % "ng-file-upload" % "5.0.9",
-          "org.webjars.bower" % "vis" % "4.7.0"
-        ).map(_.exclude("org.scalamacros", "quasiquotes_2.10")).map(_.exclude("org.scalamacros", "quasiquotes_2.10.3"))
-      )
-  ) dependsOn(streaming % "test->test; compile->compile", daemon % "test->test; compile->compile")
+  lazy val services = Project(id = "gearpump-services", base = file("services")).
+    settings(commonSettings: _*).
+    settings(compile in Compile <<= (compile in Compile) dependsOn (compile in (serviceJVM, Compile))).
+    aggregate(serviceJS, serviceJVM).
+    dependsOn(serviceJVM % "test->test;compile->compile", 
+      streaming % "test->test;compile->compile", 
+      daemon % "test->test;compile->compile"
+    )
+
+  lazy val service = crossProject.in(file("services")).
+    settings(commonSettings: _*).
+    jvmSettings(jvmSettings: _*). 
+    jsSettings(jsSettings: _*)
+
+  lazy val js2jvmSettings = Seq(fastOptJS, fullOptJS, packageJSDependencies) map { packageJSKey =>
+    crossTarget in(serviceJS, Compile, packageJSKey) := file(distDashboardDirectory)
+  }
+
+  lazy val serviceJVM: Project = service.jvm.
+    settings(js2jvmSettings: _*).
+    settings(compile in Compile <<= (compile in Compile) dependsOn (fastOptJS in (serviceJS, Compile))).
```

Comment:
To my understanding, crossProject is a util which help us to define the source folder of each project.

Take following example:
```
name := "Foo root project"

lazy val root = project.in(file(".")).
  aggregate(fooJS, fooJVM).
  settings(
    publish := {},
    publishLocal := {}
  )

lazy val foo = crossProject.in(file(".")).
  settings(
    name := "foo",
    version := "0.1-SNAPSHOT",
    scalaVersion := "2.11.5"
  ).
  jvmSettings(
    // Add JVM-specific settings here
  ).
  jsSettings(
    // Add JS-specific settings here
  )

lazy val fooJVM = foo.jvm
lazy val fooJS = foo.js
```

What is meanful is project fooJvm and fooJs, project foo is only a Util to generate settings for fooJvm and fooJs.

If we respect this rule, then there are several suggestions:
1). for project lazy val services, since now it only becomes a util, it should be referenced anywhere in Build.sbt. We should only reference serviceJs or serviceJVm.

So the following code doesn't make sense to me
```
  lazy val services = Project(id = "gearpump-services", base = file("services")).
    aggregate(serviceJS, serviceJVM).

// All lines below doesn't make sense to exist.
    settings(commonSettings: _*).
    settings(compile in Compile <<= (compile in Compile) dependsOn (compile in (serviceJVM, Compile))).
    dependsOn(serviceJVM % "test->test;compile->compile", 
      streaming % "test->test;compile->compile", 
      daemon % "test->test;compile->compile"
    )
```